BlueprintUICommonControls/Sources/LifecycleObserver.swift

```diff
@@ -0,0 +1,173 @@
+        self.onMount = onMount
+        self.onUnmount = onUnmount
+        self.wrapped = wrapping
+    }
+
+    public var content: ElementContent {
+        ElementContent(child: wrapped)
+    }
+
+    public func backingViewDescription(with context: ViewDescriptionContext) -> ViewDescription? {
+        PassthroughView.describe { (config) in
+            config.onAppear = onAppear
+            config.onDisappear = onDisappear
+            config.onMount = onMount
+            config.onUnmount = onUnmount
+        }
+    }
+}
+
+// These extensions collapse chained callbacks into a single element
+extension LifecycleObserver {
+    /// Adds a hook that will be called when this element appears.
+    /// 
+    /// For more details about lifecycle hooks, see `LifecycleCallback`.
+    ///
+    /// ## In Xcode
+    /// [The Element Lifecycle](x-source-tag://ElementLifecycle)
+    /// 
+    public func onAppear(_ callback: @escaping LifecycleCallback) -> LifecycleObserver {
+        LifecycleObserver(
+            onAppear: onAppear + callback,
+            onDisappear: onDisappear,
+            onMount: onMount,
+            onUnmount: onUnmount,
+            wrapping: wrapped
+        )
+    }
+
+    /// Adds a hook that will be called when this element disappears.
+    /// 
+    /// For more details about lifecycle hooks, see `LifecycleCallback`.
+    ///
+    /// ## In Xcode
+    /// [The Element Lifecycle](x-source-tag://ElementLifecycle)
+    /// 
+    public func onDisappear(_ callback: @escaping LifecycleCallback) -> LifecycleObserver {
+        LifecycleObserver(
+            onAppear: onAppear,
+            onDisappear: onDisappear + callback,
+            onMount: onMount,
+            onUnmount: onUnmount,
+            wrapping: wrapped
+        )
+    }
+
+    /// Adds a hook that will be called when this element is mounted.
+    /// 
+    /// For more details about lifecycle hooks, see `LifecycleCallback`.
+    ///
+    /// ## In Xcode
+    /// [The Element Lifecycle](x-source-tag://ElementLifecycle)
+    /// 
+    public func onMount(_ callback: @escaping LifecycleCallback) -> LifecycleObserver {
+        LifecycleObserver(
+            onAppear: onAppear,
+            onDisappear: onDisappear,
+            onMount: onMount + callback,
+            onUnmount: onUnmount,
+            wrapping: wrapped
+        )
+    }
+
+    /// Adds a hook that will be called when this element is unmounted.
+    /// 
+    /// For more details about lifecycle hooks, see `LifecycleCallback`.
+    ///
+    /// ## In Xcode
+    /// [The Element Lifecycle](x-source-tag://ElementLifecycle)
+    /// 
+    public func onUnmount(_ callback: @escaping LifecycleCallback) -> LifecycleObserver {
+        LifecycleObserver(
+            onAppear: onAppear,
+            onDisappear: onDisappear,
+            onMount: onMount,
+            onUnmount: onUnmount + callback,
+            wrapping: wrapped
+        )
+    }
+}
+
+extension Element {
```

Comment:
Might be also nice to add this version?

```
func observe(
        onAppear: LifecycleCallback? = nil,
        onDisappear: LifecycleCallback? = nil,
        onMount: LifecycleCallback? = nil,
        onUnmount: LifecycleCallback? = nil,
) -> LifecycleObserver
```

Reply:
or
```swift
func on(
    mount: LifecycleCallback? = nil,
    appear: LifecycleCallback? = nil,
    disappear: LifecycleCallback? = nil,
    unmount: LifecycleCallback? = nil
) -> LifecycleObserver
```

@kyleve are you suggesting this from a performance perspective (less nested elements)?

Reply:
Because of the extensions on `LifecycleObserver`, calling

```swift
element
  .onAppear { }
  .onDisappear { }
  .onMount { }
  .onUnmount { }
```

Will coalesce to a single `LifecycleObserver`.

Reply:
Mostly just since it makes the element tree cleaner (especially if debugging); but nesting would have a non-zero cost as well.

Reply:
(Oops, just saw Andrew and I replied at the same time)